proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,482 @@
+
+Data structure re-use
+~~~~~~~~~~~~~~~~~~~~~
+
+With levity polymorphism, we can even re-use currently lifted-only data structures:
+
+::
+  
+  data List a :: TYPE (BoxedRep l) where
+    Nil :: List a
+    Cons :: a -> List a -> List a
+
+  mapLifted :: (a -> b) -> List a -> List b
+  mapLifted f Nil         = Nil
+  mapLifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
+  
+  mapUnlifted :: (a -> b) -> List @Unlifted a -> List @Unlifted b
+  mapUnlifted f Nil         = Nil
+  mapUnlifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
+  
+There no chance of sharing the ``map`` definition (not this one anyway)
+currently, because we don't have levity polymorphism in expressions yet, which
+should be tackled in a separate proposal.
+
+Effect and Interactions
+-----------------------
+
+Introduction of user-defined unlifted data types means we can finally write
+code processing data types that can be compiled as if we were in a strict
+language.
+
+**Strict constructor fields** share considerable overlap with ``Strict``, yet they
+proved insufficient for encoding invariants for efficient code generation.
+
+This proposal consciously left out further work like a new specification for
+**levity polymorphic code**. Similar to data types, functions can be levity
+polymorphic, too. There's
+`#15532 <https://gitlab.haskell.org/ghc/ghc/issues/15532>`_,
+which wants to weaken the restrictions we have in place for runtime-rep
+polymorphic code.
+
+**Pattern match checking** with unlifted types will be weird in some edge cases.
+Consider the following example:
+
+::
+ 
+  data SVoid :: TYPE (BoxedRep Unlifted)
+  f :: SVoid -> ()
+  f _ = ()
+
+Should this program be accepted without warning? It would be accepted for
+lifted ``Void``, because ``f (error "boom")`` is a valid call and would
+evaluate to ``()``. But with unlifted ``Void`` this doesn't make sense anymore:
+Because of call-by-value, the ``error`` thunk will be evaluated before entering
+``f``, resulting in a crash. In that regard, it's similar to the situation with
+
+::
+
+  data Void
+  f :: Void -> ()
+  f !_ = ()
+
+And should probably elicit an inaccessible RHS warning. I guess this is accurate
+for unlifted functions as well as long as we don't allow functions without
+bindings.
+
+**-XStrict/-XStrictData** could implicitly turn all data declarations
+into unlifted ones. I see two potential problems:
+
+* If a data type is exported, it's now an unlifted type. That's a breaking change.
+* For data family instances, this is only possible if the parent data family
+  was kind polymorphic. Plus it's a strange thing to do change kinds of a
+  declaration just by switching on a language extension.
+
+So rather dreadfully, we probably shouldn't "augment" ``-XStrict``.
+
+**Lazy pattern matches** ``let ~t = ... in ...`` should not be allowed if ``t``
+is unlifted. That is exactly the behavior that is currently implemented in GHC.
+
+**Data con wrappers** of nullary constructors like
```

Comment:
This stuff about data con wrappers is *only* about GHC's implementation, nothing to do with the specification.  Please say so.   Examples woudl be useful.

Reply:
Added an example and clarified that this is about the implementation.